test/unit/utils/URITemplate.ts

```diff
@@ -0,0 +1,190 @@
+import chai from "chai";
+import chaiAsPromised from "chai-as-promised";
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+import { ethers } from "hardhat";
+import { TestURITemplate, TestURITemplateLib } from "../../../typechain-types";
+
+chai.use(chaiAsPromised);
+const expect = chai.expect;
+
+let contract: TestURITemplate;
+let lib: TestURITemplateLib;
+let accounts: SignerWithAddress[];
+
+describe("URITemplate", function () {
+  // Deploy the contract and get signers
+  beforeEach(async function () {
+    const UriTemplate = await ethers.getContractFactory("TestURITemplate");
+    contract = (await UriTemplate.deploy()) as TestURITemplate;
+    await contract.deployed();
+    accounts = await ethers.getSigners();
+  });
+
+  it("Should not create invalid template URI", async function () {
+    // Check that empty strings do not set a new URI
+    await expect(contract.setURITemplateSingle("")).to.be.rejectedWith(
+      "InvalidTemplate()"
+    );
+    await expect(contract.setURITemplateMultiple("")).to.be.rejectedWith(
+      "InvalidTemplate()"
+    );
+  });
+
+  it("Should return empty strings for unset token URI", async function () {
+    // Mint an NFT
+    const minter = accounts[1];
+    const tx = await contract.connect(minter).mint();
+    const receipt = await tx.wait();
+    const [event] = receipt.events ?? [];
+    const tokenId = event.args?.tokenId;
+    // Check that empty strings are returned
+    expect(await contract.tokenURI(tokenId)).to.equal(``);
+  });
+
+  it("Should create template URI with single 'id'", async function () {
+    // Set token URI
+    await contract.setURITemplateSingle("https://foo.xyz/{id}/bar");
+    // Mint an NFT
+    const minter = accounts[1];
+    const tx = await contract.connect(minter).mint();
+    const receipt = await tx.wait();
+    const [event] = receipt.events ?? [];
+    const tokenId = event.args?.tokenId;
+    expect(await contract.tokenURI(tokenId)).to.equal(
+      `https://foo.xyz/${tokenId}/bar`
+    );
+
+    await contract.setURITemplateMultiple("https://foo.xyz/{id}/asdf");
+    expect(await contract.tokenURI(tokenId)).to.equal(
+      `https://foo.xyz/${tokenId}/asdf`
+    );
+  });
+
+  it("Should fail to create URI with multiple 'id' using the 'single' method", async function () {
+    // Set token URI
+    await contract.setURITemplateSingle("https://foo.xyz/{id}/bar={id}/foo");
+    // Mint an NFT
+    const minter = accounts[1];
+    const tx = await contract.connect(minter).mint();
+    const receipt = await tx.wait();
+    const [event] = receipt.events ?? [];
+    const tokenId = event.args?.tokenId;
+    // Check the token URI is the incorrect format
+    expect(await contract.tokenURI(tokenId)).to.equal(
+      `https://foo.xyz/${tokenId}/bar={id}/foo`
+    );
+  });
```

Comment:
Is there any reasonable way to make `contract.setURITemplateSingle` error/revert in this case? Seems strange to let it all succeed and end up with results you didn't expect.

Reply:
@asutula totally, I'll add this! I initially didn't want to over engineer or write all of the tests until I was certain which approach I was taking, which was short sighted.